### PR TITLE
Runner - Implement ansible client and runner

### PR DESCRIPTION
**Pull request by abuccts**
__Major Revisions__
* add ansible client
* add deploy and check_env playbooks

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -104,6 +104,7 @@ __Install SuperBench__
 
     # install superbench
     python3 -m pip install .
+    ansible-galaxy collection install ansible.utils community.crypto
```
[the remark this answers is not in the document]
Comment (abuccts) [reply README.md:107]: it only works in develop mode, like `python setup.py develop` or `pip install -e`.
`pip install` will build dist wheel and doesn't allow this kind of arbitrary code execution for safety by design.

File: .azure-pipelines/cpu-unit-test.yml [hunk 2]
```diff
@@ -8,8 +8,12 @@ container:
   image: python:3.7
 
 steps:
+  - script: |
+      echo "##vso[task.prependpath]$HOME/.local/bin"
+    displayName: Export path
   - script: |
       python3 -m pip install .[test,torch]
+      ansible-galaxy collection install ansible.utils community.crypto
```
[the remark this answers is not in the document]
Comment (abuccts) [reply .azure-pipelines/cpu-unit-test.yml:16]: same

File: superbench/runner/runner.py [hunk 3]
```diff
@@ -3,9 +3,13 @@
```
[the remark this answers is not in the document]
Comment (abuccts) [reply superbench/runner/runner.py:3]: done

File: README.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (cp5555) [reply README.md:107]: Why not `pip install -e` in the installation directly?

File: .azure-pipelines/cpu-unit-test.yml (hunk 2 above)
[the remark this answers is not in the document]
Comment (cp5555) [reply .azure-pipelines/cpu-unit-test.yml:16]: Why not use `pip install -e` directly?

**Comment by cp5555** [comment]
Change the description.

File: superbench/runner/ansible.py [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+        }
+        if config:
+            inventory_file = getattr(config, 'host_file', None)
+            if inventory_file:
+                self._config['inventory'] = inventory_file
+                self._config['host_pattern'] = 'all'
+                inventory = InventoryManager(loader=DataLoader(), sources=inventory_file)
+                host_list = inventory.get_groups_dict()['all']
+                if len(host_list) > 128:
+                    self._config['cmdline'] = '--forks {}'.format(len(host_list))
```
Comment (cp5555) [inline superbench/runner/ansible.py:38]: Why check the length of `host_list`?

File: superbench/runner/ansible.py [hunk 5]
```diff
@@ -0,0 +1,90 @@
+# Copyright (c) Microsoft Corporation.
```
Comment (cp5555) [inline superbench/runner/ansible.py:1]: Add UT for this class `AnsibleClient`.

File: superbench/runner/playbooks/check_env.yaml [hunk 6]
```diff
@@ -0,0 +1,41 @@
+- name: Runtime Environment Check
```
Comment (cp5555) [inline superbench/runner/playbooks/check_env.yaml:1]: When Runtime Environment Check fails, does ansible run Runtime Environment Update?

Comment (cp5555) [inline superbench/runner/playbooks/check_env.yaml:1]: We should write UT for each playbook.

File: superbench/runner/runner.py [hunk 7]
```diff
@@ -3,9 +3,13 @@
 
 """SuperBench Runner."""
```
Comment (cp5555) [inline superbench/runner/runner.py:4]: After reading the code, this integration should be separated from this PR. We should write UT to make sure class `AnsibleClient` and related playbook work. After this PR merged, we can create PR for this integration.

Comment (cp5555) [reply superbench/runner/runner.py:4]: After discussion, we can add PR for UT separately.

**Review by cp5555 — APPROVED** [review]

File: setup.py [hunk 8]
```diff
@@ -103,6 +104,15 @@ def run(self):
         sys.exit(0 if errno == 0 else 1)
 
 
+class PostDevelop(develop):
+    """Post-installation for development mode."""
+    def run(self):
+        """Override develop.run."""
+        develop.run(self)
+        errno = os.system('ansible-galaxy collection install ansible.utils community.crypto')
```
Comment (guoshzhao) [inline setup.py:112]: This command is also needed for production, right? So why call it develop mode? It looks a little ambiguous.

File: README.md (hunk 1 above)
Comment (guoshzhao) [inline README.md:107]: If the command is necessary for SB, I think we should package the command, just as you said python setup.py xxxx

File: superbench/runner/ansible.py [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+                }
+            key_file = getattr(config, 'private_key', None)
+            if key_file:
+                self._config['cmdline'] += ' --private-key {}'.format(key_file)
+            elif password:
+                self._config['cmdline'] += ' --ask-pass --ask-become-pass'
+        logger.info(self._config)
+
+    def run(self, ansible_config, sudo=False):
+        """Run Ansible runner.
```
Comment (guoshzhao) [inline superbench/runner/ansible.py:56]: Args description.

File: superbench/runner/ansible.py [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+            ansible_config['cmdline'] += ' --become'
+        r = ansible_runner.run(**ansible_config)
+        if r.rc == 0:
+            logger.info('Run succeed, return code {}.'.format(r.rc))
+        else:
+            logger.warning('Run failed, return code {}.'.format(r.rc))
+        logger.info(r.stats)
+        return r.rc
+
+    def run_shell(self, cmd, sudo=False):
```
Comment (guoshzhao) [inline superbench/runner/ansible.py:72]: Needn't to check the return flag?

File: superbench/runner/ansible.py [hunk 11]
```diff
@@ -0,0 +1,90 @@
...
+
+    def run_playbook(self, playbook, extravars=None):
+        """Run playbook in Ansible."""
+        logger.info('Run playbook {} ...'.format(playbook))
+        ansible_config = {
+            **self._config,
+            'extravars': extravars,
+            'playbook': str(self._playbook_path / playbook),
+        }
+        self.run(ansible_config)
```
Comment (guoshzhao) [inline superbench/runner/ansible.py:90]: Needn't to check the return flag?

File: superbench/runner/runner.py [hunk 12]
```diff
@@ -23,11 +27,16 @@ def __init__(self, sb_config, docker_config, ansible_config, output_dir):
...
         self._ansible_config = ansible_config
         self._output_dir = output_dir
+        self._ansible_client = AnsibleClient(ansible_config)
 
         self.__set_logger('sb-run.log')
         logger.info('Runner uses config: %s.', self._sb_config)
         logger.info('Runner writes to: %s.', self._output_dir)
 
+        self._sb_benchmarks = self._sb_config.superbench.benchmarks
+        self._sb_enabled = self.__get_enabled_benchmarks()
```
Comment (guoshzhao) [inline superbench/runner/runner.py:37]: _sb_enabled looks like a bool variable.

File: README.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (abuccts) [reply README.md:107]: changed to `make postinstall`

File: .azure-pipelines/cpu-unit-test.yml (hunk 2 above)
[the remark this answers is not in the document]
Comment (abuccts) [reply .azure-pipelines/cpu-unit-test.yml:16]: updated

File: superbench/runner/ansible.py (hunk 4 above)
Comment (abuccts) [reply superbench/runner/ansible.py:38]: check `len > 0` instead

File: superbench/runner/playbooks/check_env.yaml (hunk 6 above)
Comment (abuccts) [reply superbench/runner/playbooks/check_env.yaml:1]: no, it will exit at fail

File: setup.py (hunk 8 above)
Comment (abuccts) [reply setup.py:112]: move to `make postinstall`

File: README.md (hunk 1 above)
Comment (abuccts) [reply README.md:107]: use `make postinstall` instead

File: superbench/runner/ansible.py (hunk 5 above)
Comment (abuccts) [reply superbench/runner/ansible.py:1]: added

File: superbench/runner/playbooks/check_env.yaml (hunk 6 above)
Comment (abuccts) [reply superbench/runner/playbooks/check_env.yaml:1]: will add in separate PR

File: superbench/runner/runner.py [hunk 13]
```diff
@@ -58,7 +58,7 @@ def __get_enabled_benchmarks(self):
                 return list(self._sb_config.superbench.enable)
         return [k for k, v in self._sb_benchmarks.items() if v.enable]
 
-    def deploy(self):
+    def deploy(self):    # pragma: no cover
```
Comment (cp5555) [inline superbench/runner/runner.py:61]: What's the meaning of `# pragma: no cover`?

**Review by guoshzhao — APPROVED** [review]

File: superbench/runner/runner.py (hunk 13 above)
Comment (abuccts) [reply superbench/runner/runner.py:61]: it is used to skip coverage calculation in unit tests for specific lines/functions
these functions will be covered in ansible integration tests

**Pull request merged**